src/mds/MDCache.cc

```diff
@@ -11731,101 +11734,235 @@ void MDCache::scrub_dentry_work(MDRequestRef& mdr)
   if (!locked)
     return;
 
-  CInode::validated_data *vr =
-      static_cast<CInode::validated_data*>(mdr->internal_op_private);
+  C_MDS_EnqueueScrub *cs = static_cast<C_MDS_EnqueueScrub*>(mdr->internal_op_finish);
+  ScrubHeaderRef &header = cs->header;
+
+  // Cannot scrub same dentry twice at same time
+  if (in->scrub_info()->scrub_in_progress) {
+    mds->server->respond_to_request(mdr, -EBUSY);
+    return;
+  }
+
+  header->origin = in;
 
-  in->validate_disk_state(vr, mdr, NULL);
+  // only set completion context for non-recursive scrub, because we don't 
+  // want to block asok caller on long running scrub
+  Context *fin = NULL;
+  if (!header->recursive)
+    fin = cs->take_finisher();
+
+  mds->scrubstack->enqueue_inode_bottom(in, header, fin);
```

Comment:
If this is a non-recursive scrub which we're blocking the admin socket for, we probably want to put it on top of the stack for earliest service, don't we?